### DISCUSSION TITLE
CPU count on macOS

### DISCUSSION THREAD

**Discussion by ZRunner** [sb-ai-lab/Eco2AI#4] [Q&A] [answered]
Hello there!

I just discovered this lib and am looking to use it in one of my Python projects, but during my recent tests I noticed something intriguing when looking up the number of CPUs on a mac (Darwin system, in this case a Mac Mini M1).

From what I understand, you are using the "hw.cpu64bit_capable" result of the `sysctl` command to determine the number of available CPUs. However, from what I could find in the few documentations, this result is a boolean (0 or 1) indicating if the CPU supports 64 bits programs.  
So, is it really the right criterion to know the number of CPU? Does a Mac necessarily have only one physical CPU?

To give a little more context to my thinking, here are the results of the `sysctl -a| sort | grep cpu` command on my mac:
```
hw.activecpu: 8
hw.cpu64bit_capable: 1
hw.cpufamily: 458787763
hw.cpusubfamily: 2
hw.cpusubtype: 2
hw.cputype: 16777228
hw.logicalcpu: 8
hw.logicalcpu_max: 8
hw.ncpu: 8
hw.perflevel0.cpusperl2: 4
hw.perflevel0.logicalcpu: 4
hw.perflevel0.logicalcpu_max: 4
hw.perflevel0.physicalcpu: 4
hw.perflevel0.physicalcpu_max: 4
hw.perflevel1.cpusperl2: 4
hw.perflevel1.logicalcpu: 4
hw.perflevel1.logicalcpu_max: 4
hw.perflevel1.physicalcpu: 4
hw.perflevel1.physicalcpu_max: 4
hw.physicalcpu: 8
hw.physicalcpu_max: 8
kern.cpu_checkin_interval: 4000
kern.hv.hc_vcpu_kick: 1
kern.hv.hc_vcpu_wfk: 1
kern.sched_rt_avoid_cpu0: 0
machdep.cpu.brand_string: Apple M1
machdep.cpu.core_count: 8
machdep.cpu.cores_per_package: 8
machdep.cpu.logical_per_package: 8
machdep.cpu.thread_count: 8
```

Thanks in advance!

**Comment by vladimir-laz** [comment, marked as answer]
Hello, [ZRunner](https://github.com/ZRunner)! 

Thanks for your question!
Sorry for not answering for a long time. We had to check eco2AI functionality mentioned in your question.
As for using "hw.cpu64bit_capable" - you are completely right, it came from early versions and now(thanks to you) we fixed it. In newer versions of eco2AI number of physical CPUs for MacOS will be set to 1, because the latest models(starting from 2013) have only one physical CPU. 
So thank you for noticing a possible error. For now, almost all CPUs support 64 bits programs, so we don’t expect any problems with current version of the lib(with your specific Mac system as well). 

“Does a Mac necessarily have only one physical CPU?” - As it was mentioned above, yes.  

In addition, if you have any idea of how to improve our library, we will be happy to discuss it. 
I hope, I’ve managed to answer your question. If not, please, specify more details

Best wishes, 
Vladimir
